Add initial support for HIFI4 DSP on imxrt600 board

dbaluta opened the pull request:
* this adds SOC definitions
* DTS
* board definition

dbaluta committed:
soc: imxrt6xx: Introduce initial SoC support for HIFI4 DSP

Add SoC definitions and a linker file for the HIFI4 DSP
found on the NXP i.MXRT685 microcontroller.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>
Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>

dbaluta committed:
boards: mimxrt685_evk: Add initial board support HIFI4 DSP

This adds a board for HIFI4 DSP found on i.MXRT865 microcontroller.
It adds the initial configuration files, dts and documentation.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>
Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>

dbaluta commented:
Changes since v1:
* fix whitespaces to make compliance check happy

dbaluta commented:
Changes since v2:
- removed west sign

LaurentiuM1234 commented on soc/nxp/imxrt/imxrt6xx/Kconfig.defconfig:
```diff
@@ -115,3 +115,32 @@ choice USB_MCUX_CONTROLLER_TYPE
+config CACHE
+	default n
+
+config DCACHE
```
This should be already disabled if `CONFIG_CPU_HAS_DCACHE` is disabled, which seems to be the case here

dbaluta replied:
True. will remove it.

LaurentiuM1234 commented on soc/nxp/imxrt/imxrt6xx/Kconfig.defconfig:
```diff
@@ -115,3 +115,32 @@ choice USB_MCUX_CONTROLLER_TYPE
+config DCACHE
+	default n
+
+config CACHE_MANAGEMENT
```
What's this for? Isn't this disabled by default?

dbaluta replied:
Yes, this depends and it is disabled by default.

LaurentiuM1234 commented on soc/nxp/imxrt/imxrt6xx/Kconfig.defconfig:
```diff
@@ -115,3 +115,32 @@ choice USB_MCUX_CONTROLLER_TYPE
+config CACHE_MANAGEMENT
+	default n
+
+config LOG
```
Same here. Isn't this disabled by default?

dbaluta replied:
Yes. Thi sis defined like this:

```

config LOG                                                                                                                                                                                                         
»       bool "Logging"
»       select PRINTK if USERSPACE
»       help
»         Global switch for the logger, when turned off log calls will not be
»         compiled in.
```

and looks like if default value is not present than is `N`.

LaurentiuM1234 commented on soc/nxp/imxrt/imxrt6xx/Kconfig.defconfig:
```diff
@@ -115,3 +115,32 @@ choice USB_MCUX_CONTROLLER_TYPE
+config SYS_CLOCK_TICKS_PER_SEC
+	default 1000
+
+config CACHE
```
What's this for? I wasn't able to find `CONFIG_CACHE` anywhere?

dbaluta replied:
```
At drivers/cache/Kconfig:4
Included via /home/nxf25322/work/sof_dir3/zephyr/Kconfig:8 -> Kconfig.zephyr:48 -> drivers/Kconfig:14
Menu path: (Top) -> Device Drivers

  menuconfig CACHE
        bool "External cache controller drivers"
        default y if CACHE_MANAGEMENT(=y)
        help
          Enable support for external cache controllers drivers
```

I'm not sure what is the default value of a config symbol if not specified.

LaurentiuM1234 commented on soc/nxp/imxrt/imxrt6xx/adsp/include/adsp/io.h:
```diff
@@ -0,0 +1,41 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __INCLUDE_IO__
```
Is this stuff still used?

dbaluta replied:
Looks like this is a relict from SOF world. For now it is not used. Will add this back once there will be users for it! Will remove it in v3.

LaurentiuM1234 commented on soc/nxp/imxrt/imxrt6xx/adsp/include/soc.h:
```diff
@@ -0,0 +1,39 @@
+#define XTENSA_IRQ_NUMBER(_irq) \
+	((_irq >> XTENSA_IRQ_NUM_SHIFT) & XTENSA_IRQ_NUM_MASK)
+
+extern void z_soc_irq_enable(uint32_t irq);
```
What's this for?

dbaluta replied:
This looks like to be used by intc_nxp_irqsteer but imxrt685 doesn't have an irqsteer so will remove this calls.

LaurentiuM1234 commented on soc/nxp/imxrt/imxrt6xx/Kconfig.defconfig:
```diff
@@ -115,3 +115,32 @@ choice USB_MCUX_CONTROLLER_TYPE
 
 endif # SOC_MIMXRT685S_CM33
+
+if SOC_MIMXRT685S_ADSP
```
Might be worth attempting to de-centralize `Kconfig.defconfig` here (aka have a per-cluster Kconfig.deconfig placed in each cluster's directory and sourced in the main Kconfig.defconfig). Should make the main `Kconfig.defconfig` easier to read and, overall, should be cleaner. Maybe same could be applicable to `Kconfig` and `Kconfig.soc`? Although I question if it would be worth doing it for `Kconfig.soc` since it doesn't contain much?

dbaluta replied:
I think this is too much of an overkill right now and we could do that later when this files gets bigger. Also, imxrt595 already has this implementation and prefer to keep it consistent.

dbaluta commented:
Changes since v3:
- removed io.h file as this was an interface exported by Zephyr to SOF which is not yet supported
- removed default config for DCACHE, CACHE_MANAGEMENT and LOG as their default values is already correct

dleach02 commented on soc/nxp/imxrt/imxrt6xx/adsp/include/_soc_inthandlers.h:
```diff
@@ -0,0 +1,353 @@
+{
+	int irq;
+
+	if (mask & BIT(0)) {
```
Do we not have #defines for all these bit masks?

dbaluta replied:
@dleach02 this file is auto-generated by Xtensa tools and prefer to keep it like that. Same approach is taken for all intances of this file for intel_adsp, nxp_adsp, espressif.

dbaluta commented on boards/nxp/mimxrt685_evk/doc/index.rst:
```diff
@@ -350,6 +350,21 @@ steps:
+
+   $ west build -b mimxrt685_evk/mimxrt685s/adsp samples/hello_world
+
+Debugging can be directly carried out using the J-Link GDB server with ``xt-gdb`` connected. It's
```
@VitekST  can you please chime in here?

danieldegrasse commented on soc/nxp/imxrt/soc.yml:
```diff
@@ -34,3 +34,4 @@ family:
     - name: mimxrt685s
       cpuclusters:
       - name: cm33
+      - name: adsp
```
For the RT595, we name the ADSP core `f1`, since it is a fusion F1. My rational for doing this is that it follows how we name ARM core clusters in tree. Would you prefer to keep the `adsp` naming here, or refer to this DSP as the `hifi4`? I don't have a strong preference

danieldegrasse reviewed (commented):
General comment here- should we provide some form of code to boot the DSP at the SOC level? Currently, my understanding is that anyone wanting to use this enablement would need some application level code/build infrastructure to:
- build the ADSP image
- copy the ADSP image binary to the right offset in RAM (or potentially multiple offsets, if the data and text sections need to be located in different regions)
- Setup clocks for the DSP and release it from reset

I'm aware this may be out of scope for this PR, and that what we have here matches the enablement for the RT500 in tree. However, I think it would be useful for customers to have some in tree code to help booting the DSP core, and moreover would be helpful for testing purposes if we expand the ADSP enablement.

If we don't add code to do this, I think at a minimum outlining the steps required in the board documentation would be helpful

VitekST commented:
@danieldegrasse About initialisation of the DSP, I believe it's on the roadmap, as I've written a driver for exactly that purpose in https://github.com/zephyrproject-rtos/zephyr/pull/65229, which has been closed: https://github.com/zephyrproject-rtos/zephyr/pull/65229/files#diff-6916b9e13247d4852595c6b593424b9266520258663d0677a47518849078be7f

@dbaluta just took some burden off me by trying to get my code adapted to HWM2 and merged, or, at least, that's my understanding.

teburd commented:
Nice!

maxekman commented:
@dbaluta Hi! Any chance this PR can be reopened and eventually merged?

VitekST commented:
@maxekman The work is continued here: https://github.com/zephyrproject-rtos/zephyr/pull/77814

maxekman commented:
Thanks for the pointer @VitekST!
